Network/HTTP/Req.hs

```diff
@@ -937,8 +960,49 @@ uriHost uri = case URI.uriAuthority uri of
   Right URI.Authority {..} ->
     Just (URI.unRText authHost)
 
+-- | A quasiquoter to build a 'Url'. Can be used both as an expression and as a
+-- pattern. Only works for URIs that don't involve query options (e.g. port)
+--
+-- >>> [urlQ|https://exmaple.org/test|]
+-- Url Https ("test" :| ["example.org"])
+urlQ :: TH.QuasiQuoter
+urlQ = TH.QuasiQuoter
+  { quoteExp  = uriQ
+  , quotePat  = \str -> TH.appE [|(==)|] (uriQ str) `TH.viewP` [p|True|]
+  , quoteType = error "This usage is not supported"
+  , quoteDec  = error "This usage is not supported" }
+  where
+    uriQ :: String -> TH.Q TH.Exp
+    uriQ str = do
+      (url, uri) <- urlQShared str
+      case uriOptions uri of
+        Just _ -> fail "URL with options not supported, please use urlWithOptsQ"
+        Nothing -> url
+
+-- | A quasiquoter to build a 'Url' and 'Option' tuple. Can only be used as an
+-- expression.
+--
+-- >>> :t [urlWithOptsQ|https://example.org:443/test?x=1|]
+-- [urlWithOptsQ|https://exmaple.org:443/test?x=1|] :: (Url 'Https, Option scheme)
+urlWithOptsQ :: TH.QuasiQuoter
+urlWithOptsQ = TH.QuasiQuoter
+  { quoteExp  = \str -> do
+      (url, uri) <- urlQShared str
+      TH.tupE [url, [|fromMaybe mempty (uriOptions uri)|]]
+  , quotePat  = error "This usage is not supported"
+  , quoteType = error "This usage is not supported"
+  , quoteDec  = error "This usage is not supported" }
+
+urlQShared :: String -> TH.Q (TH.ExpQ, URI)
+urlQShared str = case URI.mkURI (T.pack str) of
+  Left err -> fail (displayException err)
+  Right uri -> case useURI uri of
+    Nothing -> fail "Not a URL but a local path"
+    Just eurl ->
+      pure (either (TH.lift . fst) (TH.lift . fst) eurl, uri)
+
 -- | An internal helper function to extract 'Option's from a 'URI'.
-uriOptions :: forall scheme. URI -> Option scheme
+uriOptions :: forall scheme. URI -> Maybe (Option scheme)
```

Comment:
What is the motivation behind this `Maybe`?

Reply:
Because I want to know when `uriOptions` just returns `mempty`, but `Option` has no `Eq` instance. The `Maybe` here doesn't mean that maybe it failed, I'm just using it in the  `Semigroup a => Monoid (Maybe a)` sense here. I should probably have added something to the documentation.

Reply:
If we go with a single quoter for everything the need for `Maybe` disappears. I think that's what we should do.